mesh/meshdb.go

```diff
@@ -275,43 +275,40 @@ func (m *DB) ForBlockInView(view map[types.BlockID]struct{}, layer types.LayerID
 
 // LayerBlockIds retrieves all block ids from a layer by layer index
 func (m *DB) LayerBlockIds(index types.LayerID) ([]types.BlockID, error) {
-	idsBytes, err := m.layers.Get(index.Bytes())
-	if err != nil {
-		return nil, err
-	}
-
-	if len(idsBytes) == 0 {
-		// zero block layer
-		return []types.BlockID{}, nil
+	layerBuf := index.Bytes()
+	zero := false
+	ids := []types.BlockID{}
+	it := m.layers.Find(layerBuf)
+	// TODO(dshulyak) iterator must be able to return an error
+	for it.Next() {
+		if len(it.Key()) == len(layerBuf) {
+			zero = true
+			continue
+		}
+		var id types.BlockID
+		copy(id[:], it.Key()[len(layerBuf):])
+		ids = append(ids, id)
 	}
-
-	blockIds, err := types.BytesToBlockIds(idsBytes)
-	if err != nil {
-		return nil, errors.New("could not get all blocks from database")
+	if zero || len(ids) > 0 {
+		return ids, nil
 	}
-
-	return blockIds, nil
+	return nil, database.ErrNotFound
 }
 
 // EmptyLayerHash is the layer hash for an empty layer
 var EmptyLayerHash = types.Hash32{}
 
 // AddZeroBlockLayer tags lyr as a layer without blocks
 func (m *DB) AddZeroBlockLayer(index types.LayerID) error {
-	blockIds := make([]types.BlockID, 0, 1)
-	w, err := types.BlockIdsToBytes(blockIds)
-	if err != nil {
-		return errors.New("could not encode layer blk ids")
-	}
-	err = m.layers.Put(index.Bytes(), w)
+	err := m.layers.Put(index.Bytes(), nil)
 	if err == nil {
 		m.persistLayerHash(index, EmptyLayerHash)
 	}
 	return err
 }
 
 func (m *DB) getBlockBytes(id types.BlockID) ([]byte, error) {
-	return m.blocks.Get(id.AsHash32().Bytes())
+	return m.blocks.Get(id.Bytes())
```

Comment:
do you think this should also be prefixed?

Reply:
it should be. i left it because of dependency in the fetcher, but i can change it in this pr

Reply:
a fixme works for me too